core-plugins/src/main/java/io/cdap/plugin/batch/connector/FileConnector.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    // if it is not a directory, then it is not browsable, return the path itself
+    if (!file.isDirectory()) {
+      return BrowseDetail.builder().setTotalCount(1).addEntity(generateBrowseEntity(file)).build();
+    }
+
+    // list the files and classify them with file and directory
+    File[] files = file.listFiles();
+    if (files == null) {
+      throw new IOException(String.format("Unable to browse the path %s", path));
+    }
+
+    // sort the files by names
+    Arrays.sort(files);
+    int count = 0;
+    int limit = request.getLimit() == null || request.getLimit() <= 0 ? files.length :
+                  Math.min(request.getLimit(), files.length);
+    BrowseDetail.Builder builder = BrowseDetail.builder();
+    for (int i = 0; i < files.length; i++) {
+      // do not browse hidden files
+      if (files[i].isHidden()) {
+        continue;
+      }
+
+      if (count < limit) {
+        builder.addEntity(generateBrowseEntity(files[i]));
+      }
+      count++;
+    }
+    return builder.setTotalCount(count).build();
+  }
+
+  @Override
+  protected void setConnectorSpec(ConnectorSpecRequest request, ConnectorSpec.Builder builder) {
+    super.setConnectorSpec(request, builder);
+    builder.addRelatedPlugin(new PluginSpec(FileBatchSource.NAME, BatchSource.PLUGIN_TYPE,
+                                            Collections.singletonMap("path", request.getPath())));
+  }
+
+  private BrowseEntity generateBrowseEntity(File file) throws IOException {
+    String path = file.getCanonicalPath();
+    boolean isDirectory = file.isDirectory();
+    BrowseEntity.Builder builder = BrowseEntity.builder(file.getName(), path, isDirectory ? DIRECTORY_TYPE : FILE_TYPE);
+    Set<BrowseEntityProperty> properties = new HashSet<>();
+
+    if (isDirectory) {
+      builder.canBrowse(true).canSample(true);
+    } else {
+      String fileType = FileTypeDetector.detectFileType(path);
+      builder.canSample(FileTypeDetector.isSampleable(fileType));
+      properties.add(BrowseEntityProperty.builder(FILE_TYPE_KEY, fileType,
+                                                  BrowseEntityProperty.PropertyType.STRING).build());
+      properties.add(BrowseEntityProperty.builder(SIZE_KEY, String.valueOf(file.length()),
+                                                  BrowseEntityProperty.PropertyType.SIZE_BYTES).build());
+    }
+
+    properties.add(BrowseEntityProperty.builder(LAST_MODIFIED_KEY, String.valueOf(file.lastModified()),
+                                                BrowseEntityProperty.PropertyType.TIMESTAMP_MILLIS).build());
+    properties.add(BrowseEntityProperty.builder(OWNER_KEY, Files.getOwner(file.toPath()).getName(),
+                                                BrowseEntityProperty.PropertyType.STRING).build());
+    properties.add(BrowseEntityProperty.builder(
+      GROUP_KEY, Files.readAttributes(file.toPath(), PosixFileAttributes.class).group().getName(),
```

Comment:
for non POSIX operating system, what will the value be here for group ?  will `group()` return null ?

Reply:
This plugin is supposed to only work for local file system